README
$Id: README,v 1.4 2002/07/03 05:14:04 rcaputo Exp $

--------
Abstract
--------

POE::Component::Client::HTTP is an asynchronous, event driven HTTP
user agent.  It accepts events containing HTTP::Request objects, and
it returns events carrying HTTP::Response objects.  It supports https
and simultaneous HTTP transactions.

------------
Requirements
------------

POE::Component::Client::HTTP requires http://poe.perl.org/ to be
available.  Transient network problems (or testing the module from a
disconnected machine) may cause it to fail.

This module requires POE, HTTP::Request, and HTTP::Response.  The
requirements are encoded in Makefile.PL, so the CPAN shell should do
the right thing.  If this is installed without the CPAN shell, the
person installing it will have to do the right thing manually.

SSL (https) support requires Net::SSLeay::Handle.

------------------
Basic Installation
------------------

POE::Component::Client::HTTP may be installed through the CPAN shell
in the usual CPAN shell manner.  This typically is:

  $ perl -MCPAN -e 'intstall POE::Component::Client::HTTP'

You can also read this README from the CPAN shell:

  $ perl -MCPAN -e shell
  cpan> readme POE::Component::Client::HTTP

And you can install the component from the CPAN prompt as well:

  cpan> install POE::Component::Client::HTTP

-------------------
Manual Installation
-------------------

POE::Component::Client::HTTP can also be installed manually.
<ftp://ftp.cpan.org/pub/CPAN/authors/id/R/RC/RCAPUTO/> or a similarly
named directory at your favorite CPAN mirror should hold the latest
version.

Downloading and unpacking the distribution are left as exercises for
the reader.  To build and test it:

  perl Makefile.PL
  make test

The test programs, t/01_get.t and t/02_put.t, make excellent sample
programs.  If you would like to see more details about the tests'
operations, edit them and set their DEBUG constants to any value Perl
considers "true".

When you're ready to install the component:

  make install

It should now be ready to use.

Thanks for reading!

-- Rocco Caputo / troc@netrus.net / poe.perl.org / poe.sourceforge.net


CHANGES
=========================
2004-10-02 15:37:11 v0_65
=========================

  2004-10-02 15:37:11 by rcaputo; HTTP.pm 1.58

    Bump version to 0.65 for release. 

  2004-10-01 04:04:39 by rcaputo; HTTP.pm 1.57

    David Davis discovered that HTTP response headers may be discarded if
    they are incomplete at the end of a read buffer. This commits a fix
    for his findings: Don't discard the remaining read buffer if we're
    still processing headers. 

=========================
2004-07-13 18:03:43 v0_64
=========================

  2004-07-13 18:03:43 by rcaputo; mylib/ExtUtils/AutoInstall.pm 1.2

    Upgrade the included ExtUtils::AutoInstall. 

  2004-07-13 18:02:37 by rcaputo; HTTP.pm 1.56

    Oh, duh. Increase the version number for a new release. 

  2004-07-13 18:01:56 by rcaputo; HTTP.pm 1.55

    Matt Cashner and an unidentified AOL Employee that some call... Tim
    pointed out that SSL wasn't working. "But it was for me!" I said. So
    we hunted it down to an incorrect bitwise operator in the
    set-it-blocking code for non-ActiveState Perl versions prior to
    5.8.0. This bug TOTALLY BROKE SSL for these Perl versions, just going
    to show that you can't test too much or on too many platforms. Or
    something. 

  2004-05-02 19:42:43 by rcaputo; HTTP.pm 1.54

    Scott Beck reported that the streaming code uses two different hash
    keys for the same heap member. This patch makes them consistent, and
    it resolves rt.cpan.org ticket #5810. 

  2004-05-02 19:25:25 by rcaputo; HTTP.pm 1.53

    Applied Scott Beck's patch to handle interrupted status lines. This
    commit resolves rt.cpan.org ticket #5816. 

=========================
2004-02-27 18:59:36 v0_63
=========================

  2004-02-27 18:59:36 by rcaputo; HTTP.pm 1.52; MANIFEST 1.7

    Applied Graham Barr's suggestions to (a) use Time::HiRes if it's
    available for better timeout accuracy, and (b) return an error 408
    (not 400) when a connection times out. 

=========================
2004-02-23 23:24:31 v0_62
=========================

  2004-02-23 23:24:31 by rcaputo; HTTP.pm 1.51; Makefile.PL 1.20

    Matt Cashner pointed out that HTTP.pm uses a feature from a recent
    version of POE. Makefile.PL specified the correct version for
    ExtUtils::AutoInstall, but it had the old version for
    ExtUtils::MakeMaker. This commit bumps the version to 0.62 for
    release, and it uses the same set of mandatory dependencies for both
    installer libraries. 

  2004-02-13 02:16:09 by rcaputo; HTTP.pm 1.50

    A comment and another internal test are added. 

=========================
2004-01-15 21:04:51 v0_61
=========================

  2004-01-15 21:04:51 by rcaputo; HTTP.pm 1.49

    Apply Graham Barr's patch to accept HTTP status codes without
    messages. Uses HTTP::Status to generate a default message in this
    case. 

=========================
2004-01-14 19:21:53 v0_60
=========================

  2004-01-14 19:21:53 by rcaputo; HTTP.pm 1.48

    Bump the release version to 0.60. 

  2004-01-14 19:21:27 by rcaputo; HTTP.pm 1.47

    Graham Barr discovered a problem where shutdown_output() was causing
    IIS servers to disconnect before sending responses. He suspects it's
    a race condition that only shows up over connections with
    Internet-scale latency. It didn't appear in localhost testing.
    Nevertheless, this isn't the first time where shutdown_output()
    caused problems, so I've left the call in a comment and clearly
    marked why it's that way. 

  2003-11-29 06:17:18 by rcaputo; HTTP.pm 1.46

    Apply Richard Harmon's patch, rt.cpan.org ticket 4521, to support
    "ICY" http status lines. These apparently are Shoutcast mp3 streams. 

=========================
2003-11-22 06:02:39 v0_59
=========================

  2003-11-22 06:02:39 by rcaputo; HTTP.pm 1.45; t/01_request.t 1.16

    Jeff (last name lost) discovered that redirects weren't properly
    setting the Host header for the new location. This commit fixes his
    rt.cpan.org ticket, number 3805. 

  2003-11-03 07:23:09 by rcaputo; HTTP.pm 1.44

    "Jaymes" requested that the progress indicator return raw octets
    along with the progress information. We may later want to disable the
    HTTP::Response, or to disable content building for it, to do
    progressive or streaming downloads of large things. 

=========================
2003-10-30 23:21:30 v0_57
=========================

  2003-10-30 23:21:30 by rcaputo; HTTP.pm 1.43

    Bump version for release. 

  2003-10-19 20:02:36 by rcaputo; HTTP.pm 1.42

    Add preliminary support for HTTP::HeadParser, which should better
    update base() and related headers in the HTTP::Response objects. 

  2003-10-19 20:01:55 by rcaputo; SSL.pm 1.5

    Alter the READ (and add a WRITE) function so they better follow the
    sysread() and syswrite() interfaces. 

  2003-10-19 20:01:09 by rcaputo; Makefile.PL 1.19

    Update the current versions of dependencies I have been testing with. 

  2003-10-19 16:31:10 by rcaputo
  Makefile.PL 1.18; mylib/ExtUtils/AutoInstall.pm 1.1

    Upgrade ExtUtils::AutoInstall to 0.54 

  2003-10-17 05:11:01 by rcaputo; HTTP.pm 1.41

    Alan Hasting discovered an error in the bitmap for the redirect phase
    of HTTP parsing. This should fix that error.
    
    Also we are now using delay_adjust() since it's available in POE
    0.27. This should avoid the potential for bad timeouts during long
    downloads. 

  2003-10-17 05:09:27 by rcaputo; t/01_request.t 1.15

    Remove some debugging code. 

  2003-10-17 05:08:49 by rcaputo; Makefile.PL 1.17

    We are now relying on a feature of POE 0.27. Bump up the dependency
    version number. 

=============================
Beginning of Recorded History
=============================


Request.pm
package POE::Component::Client::HTTP::Request;
use strict;
use warnings;

use POE;

use Carp;
use HTTP::Status;

# Unique request ID, independent of wheel and timer IDs.
my $request_seq = 0;

sub DEBUG () { 0 }

sub REQ_ID            () {  0 }
sub REQ_POSTBACK      () {  1 }
sub REQ_CONNECTION    () {  2 }
sub REQ_REQUEST       () {  3 }
sub REQ_STATE         () {  4 }
sub REQ_RESPONSE      () {  5 }
sub REQ_BUFFER        () {  6 }

sub REQ_OCTETS_GOT    () {  8 }
sub REQ_TIMER         () {  9 }
sub REQ_PROG_POSTBACK () { 10 }
sub REQ_USING_PROXY   () { 11 }
sub REQ_HOST          () { 12 }
sub REQ_PORT          () { 13 }
sub REQ_HISTORY       () { 14 }
sub REQ_START_TIME    () { 15 }
sub REQ_FACTORY       () { 16 }

sub RS_CONNECT      () { 0x01 }
sub RS_SENDING      () { 0x02 }
sub RS_IN_HEAD      () { 0x04 }
sub RS_REDIRECTED   () { 0x08 }
sub RS_IN_CONTENT   () { 0x20 }
sub RS_DONE         () { 0x40 }
sub RS_POSTED       () { 0x80 }

sub import {
  my ($class) = shift;

  my $package = caller();

  foreach my $tag (@_) {
    if ($tag eq ':fields') {
      foreach my $sub (qw(
	REQ_ID REQ_POSTBACK REQ_CONNECTION REQ_REQUEST REQ_STATE REQ_RESPONSE
	REQ_BUFFER REQ_OCTETS_GOT REQ_TIMER REQ_PROG_POSTBACK
	REQ_USING_PROXY REQ_HOST REQ_PORT REQ_HISTORY REQ_START_TIME
		    )) {
	no strict 'refs';
	*{$package . "::$sub"} = \&$sub;
      }
    }
    if ($tag eq ':states') {
      foreach my $sub (qw(
	RS_CONNECT RS_SENDING RS_IN_HEAD RS_REDIRECTED
	RS_IN_CONTENT RS_DONE RS_POSTED
		    )) {
	no strict 'refs';
	*{$package . "::$sub"} = \&$sub;
      }
    }
  }
}

sub ID {
  my ($self) = @_;

  return $self->[REQ_ID];
}

sub new {
  my $class = shift;

  croak __PACKAGE__ . "expects its arguments to be key/value pairs"
    if (@_ & 1);
  my %params = @_;


  croak "need a Request parameter" unless (defined $params{'Request'});
  croak "Request must be a HTTP::Request object"
      unless (UNIVERSAL::isa ($params{'Request'}, "HTTP::Request"));

  #croak "need a Tag parameter" unless (defined $params{'Tag'});

  croak "need a Factory parameter" unless (defined $params{'Factory'});

  my ($http_request, $postback, $tag, $progress, $factory) =
      @params{qw(Request Postback Tag Progress Factory)};

  my $request_id = ++$request_seq;
  DEBUG and warn "REQ: creating a request ($request_id)";

  # Get the host and port from the request object.
  my ($host, $port, $scheme, $using_proxy);

  eval {
    $host   = $http_request->uri()->host();
    $port   = $http_request->uri()->port();
    $scheme = $http_request->uri()->scheme();
  };
  croak "Not a usable Request: $@" if ($@);

  # Add a host header if one isn't included.  Must do this before
  # we reset the $host for the proxy!
  _set_host_header ($http_request)
    unless (  defined $http_request->header('Host')
	and   length $http_request->header('Host')
      );


  if (defined $params{Proxy}) {
    # This request qualifies for proxying.  Replace the host and port
    # with the proxy's host and port.  This comes after the Host:
    # header is set, so it doesn't break the request object.
    ($host, $port) = $params{Proxy}->[rand @{$params{Proxy}}];
    $using_proxy = 1;
  } else {
    $using_proxy = 0;
  }



  # Build the request.
  my $self = [
    $request_id,        # REQ_ID
    $postback,          # REQ_POSTBACK
    undef,              # REQ_CONNECTION
    $http_request,      # REQ_REQUEST
    RS_CONNECT,         # REQ_STATE
    undef,              # REQ_RESPONSE
    '',                 # REQ_BUFFER
    undef,              # unused
    0,                  # REQ_OCTETS_GOT
    undef,              # REQ_TIMER
#    "\x0D\x0A",         # REQ_NEWLINE
    $progress,          # REQ_PROG_POSTBACK
    $using_proxy,       # REQ_USING_PROXY
    $host,              # REQ_HOST
    $port,              # REQ_PORT
    undef,		# REQ_HISTORY
    time(),             # REQ_START_TIME
    $factory,           # REQ_FACTORY
   ];
   return bless $self, $class;
}

sub return_response {
  my ($self) = @_;

  DEBUG and warn "in return_response ", sprintf ("0x%02X", $self->[REQ_STATE]);
  return if ($self->[REQ_STATE] & RS_POSTED);
  my $response = $self->[REQ_RESPONSE];

  # If we have a cookie jar, have it frob our headers.  LWP rocks!
  $self->[REQ_FACTORY]->frob_cookies ($response);

  # If we're done, send back the HTTP::Response object, which
  # is filled with content if we aren't streaming, or empty
  # if we are. that there's no ARG1 lets the client know we're done
  # with the content in the latter case
  if ($self->[REQ_STATE] & RS_DONE) {
      DEBUG and warn "done; returning $response for ", $self->[REQ_ID];
      $self->[REQ_POSTBACK]->($self->[REQ_RESPONSE]);
      $self->[REQ_STATE] |= RS_POSTED;
      #warn "state is now ", $self->[REQ_STATE];
  } elsif ($self->[REQ_STATE] & RS_IN_CONTENT) {
    # If we are streaming, send the chunk back to the client session.
    # Otherwise add the new octets to the response's content.
    # This should only add up to content-length octets total!
    if ($self->[REQ_FACTORY]->is_streaming) {
      DEBUG and warn "returning partial $response";
      $self->[REQ_POSTBACK]->($self->[REQ_RESPONSE], $self->[REQ_BUFFER]);
    } else {
      DEBUG and warn "adding to $response";
      $self->[REQ_RESPONSE]->add_content($self->[REQ_BUFFER]);
    }
  }
  $self->[REQ_BUFFER] = '';
}

sub add_content {
  my ($self, $data) = @_;

  if (ref $data) {
    $self->[REQ_STATE] = RS_DONE;
    $data->scan (sub {$self->[REQ_RESPONSE]->header (@_) });
    return 1;
  }

  $self->[REQ_BUFFER] .= $data;

  # Count how many octets we've received.  -><- This may fail on
  # perl 5.8 if the input has been identified as Unicode.  Then
  # again, the C<use bytes> in Driver::SysRW may have untainted the
  # data... or it may have just changed the semantics of length()
  # therein.  If it's done the former, then we're safe.  Otherwise
  # we also need to C<use bytes>.
  # TODO: write test(s) for this.
  my $this_chunk_length = length($self->[REQ_BUFFER]);
  $self->[REQ_OCTETS_GOT] += $this_chunk_length;

  my $max = $self->[REQ_FACTORY]->max_response_size;

  DEBUG and warn  "REQ: request ", $self->ID,
		  " received $self->[REQ_OCTETS_GOT] bytes; maximum is $max";

  if (defined $max and $self->[REQ_OCTETS_GOT] > $max) {
    # We've gone over the maximum content size to return.  Chop it # back.
    my $over = $self->[REQ_OCTETS_GOT] - $max;
    $self->[REQ_OCTETS_GOT] -= $over;
    substr($self->[REQ_BUFFER], -$over) = "";
    #TODO: ??
    #$self->[REQ_STATE] |= RS_DONE;
  }

  # keep this for the progress callback (it gets cleared in return_response
  # as I say below, this needs to go away.
  my $buffer = $self->[REQ_BUFFER];

  $self->return_response;
  DEBUG and do {
    warn  "REQ: request ", $self->ID,
	  " got $this_chunk_length octets of content...";

    warn  "REQ: request ", $self->ID,
	  " has $self->[REQ_OCTETS_GOT]",
	  ( $self->[REQ_RESPONSE]->content_length()
	    ? ( " out of " . $self->[REQ_RESPONSE]->content_length() )
	    : ""
	  ),
	  " octets"
  };
  if (defined $max and $self->[REQ_OCTETS_GOT] >= $max) {
    DEBUG and
      warn  "REQ: request ", $self->ID,
	    " has a full response... moving to done.";
    $self->[REQ_STATE] |= RS_DONE;
    $self->[REQ_STATE] &= ~RS_IN_CONTENT;
    return 1;
  }

  if ($self->[REQ_RESPONSE]->content_length) {

    # Report back progress
    $self->[REQ_PROG_POSTBACK]->(
	  $self->[REQ_OCTETS_GOT],
	  $self->[REQ_RESPONSE]->content_length,
	  #TODO: ugh. this is stupid. Must remove/deprecate!
	  $buffer,
      ) if ($self->[REQ_PROG_POSTBACK]);


    # Stop reading when we have enough content.  -><- Should never be
    # greater than our content length.
    if ($self->[REQ_OCTETS_GOT] >= $self->[REQ_RESPONSE]->content_length) {
      DEBUG and
        warn  "REQ: request ", $self->ID,
	      " has a full response... moving to done.";
      $self->[REQ_STATE] |= RS_DONE;
      $self->[REQ_STATE] &= ~RS_IN_CONTENT;
      return 1;
    }
  }
  return 0;
}

sub timer {
  my ($self, $timer) = @_;

  # do it this way so we can set REQ_TIMER to undef
  if (@_ == 2) {
    $self->[REQ_TIMER] = $timer;
  }
  return $self->[REQ_TIMER];
}

sub create_timer {
  my ($self, $timeout) = @_;

  # remove old timeout first
  my $kernel = $POE::Kernel::poe_kernel;

  my $seconds = $timeout - (time() - $self->[REQ_START_TIME]);
  $self->[REQ_TIMER] = 
      $kernel->delay_set (got_timeout =>
	  $seconds, $self->ID
	);
  DEBUG and warn "TKO: request ", $self->ID,
    " has timer $self->[REQ_TIMER] going off in $seconds seconds\n";
}

sub remove_timeout {
  my ($self) = @_;

  my $alarm_id = $self->[REQ_TIMER];
  if (defined $alarm_id) {
    my $kernel = $POE::Kernel::poe_kernel;
    DEBUG and warn "REQ: Removing timer $alarm_id";
    $kernel->alarm_remove($alarm_id);
  }
}

sub postback {
  my ($self, $postback) = @_;

  if (defined $postback) {
    DEBUG and warn "REQ: modifying postback";
    $self->[REQ_POSTBACK] = $postback;
  }
  return $self->[REQ_POSTBACK];
}

sub _set_host_header {
  my ($request) = @_;
  my $uri = $request->uri;

  my ($new_host, $new_port);
  eval {
    $new_host = $uri->host();
    $new_port = $uri->port();
    # Only include the port if it's nonstandard.
    if ($new_port == 80 || $new_port == 443) {
      $request->header( Host => $new_host );
    } else {
      $request->header( Host => "$new_host:$new_port" );
    }
  };
  warn $@ if $@;
}

sub does_redirect {
  my ($self, $last) = @_;

  if (defined $last) {
    $self->[REQ_HISTORY] = $last;
    # delete OLD timeout
    #my $alarm_id = $last->[REQ_TIMEOUT];
    #DEBUG and warn "RED: Removing old timeout $alarm_id\n";
    #$POE::Kernel::poe_kernel->alarm_remove ($alarm_id);
  }

  return defined $self->[REQ_HISTORY];
}

sub check_redirect {
  my ($self) = @_;

  my $max = $self->[REQ_FACTORY]->max_redirect_count;

  if (defined $self->[REQ_HISTORY]) {
    $self->[REQ_RESPONSE]->previous($self->[REQ_HISTORY]->[REQ_RESPONSE]);
  }

  return undef unless ($self->[REQ_RESPONSE]->is_redirect);

  my $new_uri = $self->[REQ_RESPONSE]->header ('Location');
  DEBUG and warn "REQ: Preparing redirect to $new_uri";
  my $base = $self->[REQ_RESPONSE]->base();
  $new_uri = URI->new($new_uri, $base)->abs($base);
  DEBUG and warn "RED: Actual redirect uri is $new_uri";

  my $prev = $self;
  my $history = 0;
  while ($prev = $prev->[REQ_HISTORY]) {
    $history++;
    $history = $max + 1 if ($prev->[REQ_REQUEST]->uri eq $new_uri);
    last if ($history > $max);
  }

  if ($history > $max) {
    #$self->[REQ_STATE] |= RS_DONE;
    DEBUG and warn "RED: Too much redirection";
  } else { # All fine, yield new request and mark this disabled.
    my $newrequest = $self->[REQ_REQUEST]->clone();
    DEBUG and warn "RED: new request $newrequest";
    $newrequest->uri($new_uri);
    _set_host_header ($newrequest);

    $self->[REQ_STATE] = RS_REDIRECTED;
    DEBUG and warn "RED: new request $newrequest";
    return $newrequest;
  }
  return undef;
}

sub send_to_wheel {
  my ($self) = @_;

  $self->[REQ_STATE] = RS_SENDING;

  my $http_request = $self->[REQ_REQUEST];

  # MEXNIX 2002-06-01: Check for proxy.  Request query is a bit
  # different...

  my $request_uri;
  if ($self->[REQ_USING_PROXY]) {
    $request_uri = $http_request->uri()->canonical();
  }
  else {
    $request_uri = $http_request->uri()->canonical()->path_query();
  }

  my $request_string = (
    $http_request->method() . ' ' .
    $request_uri . ' ' .
    $http_request->protocol() . "\x0D\x0A" .
    $http_request->headers_as_string("\x0D\x0A") . "\x0D\x0A" .
    $http_request->content() # . "\x0D\x0A"
  );

  DEBUG and do {
    my $formatted_request_string = $request_string;
    $formatted_request_string =~ s/([^\n])$/$1\n/;
    $formatted_request_string =~ s/^/| /mg;
    warn ",----- SENDING REQUEST ", '-' x 56, "\n";
    warn $formatted_request_string;
    warn "`", '-' x 78, "\n";
  };

  $self->[REQ_CONNECTION]->wheel->put ($request_string);
}

sub wheel {
  my ($self) = @_;

  #if (defined $new_wheel) {
    # Switch wheels.  This is cumbersome, but it works around a bug in
    # older versions of POE.
    #$self->[REQ_WHEEL] = undef;
    #$self->[REQ_WHEEL] = $new_wheel;
  #}
  return $self->[REQ_CONNECTION]->wheel;
}

sub error {
  my ($self, $code, $message) = @_;

  my $nl = "\n";

  my $r = HTTP::Response->new($code);
  my $http_msg = status_message ($code);
  my $m = "<html>$nl"
      .	  "<HEAD><TITLE>Error: $http_msg</TITLE></HEAD>$nl"
      .	  "<BODY>$nl"
      .	  "<H1>Error: $http_msg</H1>$nl"
      .	  "$message$nl"
      .	  "</BODY>$nl"
      .	  "</HTML>$nl";

  $r->content ($m);
  $self->[REQ_POSTBACK]->($r);
}

sub connect_error {
  my ($self, $message) = @_;

  my $host = $self->[REQ_HOST];
  my $port = $self->[REQ_PORT];

  $message = "Cannon connect to $host:$port ($message)";
  $self->error (RC_INTERNAL_SERVER_ERROR, $message);
  return;
}

sub DESTROY {
  my ($self) = @_;

  delete $self->[REQ_CONNECTION];
  delete $self->[REQ_FACTORY];
}

1;

